### PR TITLE
[NN] add gatconv edge_weight

### PR DESCRIPTION
**Pull request by venzino-han**
## Description
related to issue #5459 

- add `edge_weight` parameter
- add test for `edge_weight` 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by jermainewang** [comment]
@dgl-bot

**Comment by mufeili** [comment]
@venzino-han I saw a "WIP" note. Is this PR ready for review?

**Comment by venzino-han** [comment]
@mufeili Not yet. I will let you know when it's ready.

**Comment by mufeili** [comment]
@venzino-han Awesome!

File: python/dgl/nn/pytorch/conv/gatconv.py [hunk 1]
```diff
@@ -327,6 +329,8 @@ def forward(self, graph, feat, get_attention=False):
             e = self.leaky_relu(graph.edata.pop("e"))
             # compute softmax
             graph.edata["a"] = self.attn_drop(edge_softmax(graph, e))
+            if edge_weight is not None:
+                graph.edata["a"] = graph.edata["a"] * edge_weight.tile(self._out_feats, self._num_heads, 1).T
```
Comment (venzino-han) [inline python/dgl/nn/pytorch/conv/gatconv.py:333]: expand edge_weigth tensor by the number of attachment heads

**Comment by venzino-han** [comment]
@mufeili It's ready for review now. Please review this when you available.

File: python/dgl/nn/pytorch/conv/gatconv.py [hunk 2]
```diff
@@ -243,6 +243,8 @@ def forward(self, graph, feat, get_attention=False):
             :math:`D_{in}` is size of input feature, :math:`N` is the number of nodes.
             If a pair of torch.Tensor is given, the pair must contain two tensors of shape
             :math:`(N_{in}, *, D_{in_{src}})` and :math:`(N_{out}, *, D_{in_{dst}})`.
+        edge_weight : torch.Tensor, optional
+            An 1D tensor of edge weight value.  Shape: :math:`(|E|,)`.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatconv.py:247]:
```suggestion
            A 1D tensor of edge weight values.  Shape: :math:`(|E|,)`.
```
?

File: python/dgl/nn/pytorch/conv/gatconv.py [hunk 3]
```diff
@@ -327,6 +329,11 @@ def forward(self, graph, feat, get_attention=False):
             e = self.leaky_relu(graph.edata.pop("e"))
             # compute softmax
             graph.edata["a"] = self.attn_drop(edge_softmax(graph, e))
+            if edge_weight is not None:
+                graph.edata["a"] = (
+                    graph.edata["a"]
+                    * edge_weight.tile(self._out_feats, self._num_heads, 1).T
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/gatconv.py:335]: `graph.edata["a"]` is of shape `(num_edges, num_heads, 1)`. I don't think `edge_weight.tile(self._out_feats, self._num_heads, 1).T` is right.

File: tests/python/pytorch/nn/test_nn.py [hunk 4]
```diff
@@ -579,6 +579,27 @@ def test_gat_conv_bi(g, idtype, out_dim, num_heads):
     assert a.shape == (g.num_edges(), num_heads, 1)
 
 
+@parametrize_idtype
+@pytest.mark.parametrize("g", get_cases(["bipartite"], exclude=["zero-degree"]))
+@pytest.mark.parametrize("out_dim", [1, 2])
+@pytest.mark.parametrize("num_heads", [1, 4])
+def test_gat_conv_edge_weight(g, idtype, out_dim, num_heads):
+    g = g.astype(idtype).to(F.ctx())
+    ctx = F.ctx()
```
Comment (mufeili) [inline tests/python/pytorch/nn/test_nn.py:588]: Interchange L588 and 587.

File: python/dgl/nn/pytorch/conv/gatconv.py [hunk 5]
```diff
@@ -327,6 +329,11 @@ def forward(self, graph, feat, get_attention=False):
             e = self.leaky_relu(graph.edata.pop("e"))
             # compute softmax
             graph.edata["a"] = self.attn_drop(edge_softmax(graph, e))
+            if edge_weight is not None:
+                graph.edata["a"] = (
+                    graph.edata["a"]
+                    * edge_weight.tile(1, self._num_heads, 1).transpose(0, 2)
```
Comment (venzino-han) [inline python/dgl/nn/pytorch/conv/gatconv.py:335]: match shape with attention tensor

File: tests/python/pytorch/nn/test_nn.py (hunk 4 above)
Comment (venzino-han) [reply tests/python/pytorch/nn/test_nn.py:587]: I also changed existing test cases

File: python/dgl/nn/pytorch/conv/gatconv.py (hunk 3 above)
Comment (venzino-han) [reply python/dgl/nn/pytorch/conv/gatconv.py]: It was my bad. I fixed the code and confirm that attention vector and edge_weight vector are same shape.

**Comment by mufeili** [comment]
@dgl-bot

**Comment by mufeili** [comment]
You also need to fix the lint issues reported. Click `Details` associated with `Lint / lintrunner (pull_request)` to see the issues.

File: python/dgl/nn/pytorch/conv/gatconv.py (hunk 3 above)
Comment (mufeili) [reply python/dgl/nn/pytorch/conv/gatconv.py]: Thank you. This works now.

**Comment by mufeili** [comment]
Some unit tests failed. See the error messages below.

```
shape = 8

    def randn(shape):
>       return th.randn(*shape)
E       TypeError: randn() argument after * must be an iterable, not int

tests\backend\pytorch\__init__.py:23: TypeError
__________________ test_gat_conv_edge_weight[4-2-g1-idtype1] __________________

g = Graph(num_nodes={'_U': 2, '_V': 4},
      num_edges={('_U', '_E', '_V'): 8},
      metagraph=[('_U', '_V', '_E')])
idtype = torch.int64, out_dim = 2, num_heads = 4

    @parametrize_idtype
    @pytest.mark.parametrize("g", get_cases(["bipartite"], exclude=["zero-degree"]))
    @pytest.mark.parametrize("out_dim", [1, 2])
    @pytest.mark.parametrize("num_heads", [1, 4])
    def test_gat_conv_edge_weight(g, idtype, out_dim, num_heads):
        ctx = F.ctx()
        g = g.astype(idtype).to(ctx)
        gat = nn.GATConv(5, out_dim, num_heads)
        feat = (
            F.randn((g.number_of_src_nodes(), 5)),
            F.randn((g.number_of_dst_nodes(), 5)),
        )
        gat = gat.to(ctx)
>       ew = F.randn(g.num_edges())
```

**Comment by mufeili** [comment]
@dgl-bot

**Comment by venzino-han** [comment]
@mufeili Thank you very much for your help!

**Comment by mufeili** [comment]
> @mufeili Thank you very much for your help!

Thank you for the great efforts!

**Pull request merged**